Buffer Controller TypeScript

itsjamie opened the pull request:
### This PR will...
This adds TypeScript definitions to the buffer controller to help migrate the internals of hls.js over to TypeScript.

### Are there any points in the code the reviewer needs to double check?
I wasn't sure what controller or file should "own" the types for Segment / Track types so I created a types folder and added them there for the purpose of moving forward. If a better location or file for ownership is known can move it there.

One additional change I made was I changed the functions that were bound as event listeners to become instance methods using the `name = () => { }` definition style, rather than bind the prototype function to the instance and keep track of the reference. This is why variables such as `this.onmse` were removed and not defined.

### Checklist
- [x] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -23,26 +74,15 @@ class BufferController extends EventHandler {
-    // the value that we have set mediasource.duration to
-    // (the actual duration may be tweaked slighly by the browser)
+    
     this._msDuration = null;
```
We could initialise all these where they are declared

     private _msDuration: number | null = null

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -149,30 +186,21 @@ class BufferController extends EventHandler {
   }
 
   checkPendingTracks () {
+    // here she be
```
Typo

itsjamie replied:
Removed it.

tjenkinson commented on src/types/segment.ts:
```diff
@@ -0,0 +1,6 @@
+export interface Segment { 
+  type: string;
+  data: any;
```
Can we not use a type here? ArrayBuffer maybe?

itsjamie replied:
I wasn't sure what the type was of data, I can take a second look to see.

itsjamie replied:
ArrayBuffer works.

tjenkinson commented on src/types/track.ts:
```diff
@@ -0,0 +1,27 @@
+  initSegment: Uint8Array;
+  levelCodec: string;
+  metadata: {
+      channelCount: any;
```
Type here? It would be great to stay away from 'any'

If it's optional


     channelCount?: number

itsjamie replied:
I'd rather remove the two metadata defintions as they aren't used in the buffer controller and it be easier to correctly define from the caller side. This was just the definition I copied from the implicit one VSCode had created.

So when the file that triggered the event that came into buffer-controller is converted, this can be extended.

tjenkinson commented on src/types/track.ts:
```diff
@@ -0,0 +1,27 @@
+  initSegment: Uint8Array;
+  levelCodec: string;
+  metadata: {
+      width: any;
```
'any'?

itsjamie replied:
Same as above. Will remove the metadata key.

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -1,16 +1,67 @@
+  
+  public tracks: TrackSet;
+  public pendingTracks: TrackSet;
+  public sourceBuffer: { audio: any, video: any }
```
We could use the source buffer type instead of any

itsjamie replied:
:+1:

itsjamie commented:
I updated the tooling that was doing the linting on the earlier build that caused Interface definitions to come up as `no-undef` errors. Now it prints:

```WARNING: You are currently running a version of TypeScript which is not officially supported by typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: ~3.1.1

YOUR TYPESCRIPT VERSION: 2.9.2

Please only submit bug reports when using the officially supported version.```

However, nothing seems broken so in the hope of keeping the change as small as possible, I won't upgrade the TypeScript version.

The build is still failing because of the issue in #2075. Once that is merged and on master, if we re-run Travis build here it should pass.

itsjamie commented:
Link #2070

johnBartos commented:
@itsjamie That warning is caused by esdoc: https://github.com/video-dev/hls.js/issues/2017

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -1,16 +1,79 @@
+
+// Add extension properties to SourceBuffers from the DOM API.
+declare global {
+  interface SourceBuffer {
```
'SourceBuffer' should already be provided by typescript, along with other browser APIs. Are you sure it didn't already exist?

itsjamie replied:
It does exist, but the hls.js code appends a few properties onto the SourceBuffer that don't exist normally.

So this patches those additional properties onto the type definition.

itsjamie replied:
So the SourceBuffer for this file is a combination of the SourceBuffer definition from the "dom" library, in addition to these properties, through the TypeScript "Interface Merging" they call it.

tjenkinson replied:
Ah sorry I missed that. `timestampOffset` should already be there though?

I also think this would be a good opportunity to make `ended` a class property instead of modifying the source buffer. It's better if we don't mutate globals.

Also this change actually applies globally to all files, not just this one. If we want to keep `ended` on `SourceBuffer`, then I think it would be better to create a new type like

```ts
type MySourceBuffer = SourceBuffer & {
  ended?: boolean
}
```

so that we are not changing global types

tjenkinson commented on tsconfig.json:
```diff
@@ -4,7 +4,12 @@
-    "strict": false
+    "strict": false,
+    "lib": [
+      "dom",
```
I think this might be why SourceBuffer is missing. Maybe try removing 'lib' from here?

If something breaks then we should probably include the defaults for es5 here too

itsjamie commented:
@johnBartos Ah yes, I see the message was printed already. Thanks!

michaelcunningham19 commented on src/controller/buffer-controller.ts:
```diff
@@ -1,16 +1,79 @@
+  public tracks: TrackSet = { audio: null, video: null };
+  public pendingTracks: TrackSet = { audio: null, video: null };
+  public sourceBuffer: SourceBufferSet
+  public flushRange: { start: number, end: number, type: keyof SourceBufferSet }[];
```
Could define a type for this, not sure what the best naming would be for it 😜 
Then we could use that type on `onBufferFlushing`

itsjamie replied:
Added a type `SourceBufferFlushRange`.

itsjamie commented on src/types/track.ts:
```diff
@@ -0,0 +1,20 @@
+  video: VideoTrack
+}
+
+export interface AudioTrack {
```
I'd hold off on this personally, this seems to be something that can be done if there is code that needs to refer to a MediaTrack but right now, everything gets union-typed where it doesn't care what type it is (AudioTrack vs VideoTrack).

Example case of where this happens is:

`let track = tracks[trackName as keyof TrackSet];`

Track is defined as `AudioTrack | VideoTrack`

michaelcunningham19 replied:
Cool 👍

itsjamie commented:
Blocked by #2075.

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -1,16 +1,82 @@
+
+// Add extension properties to SourceBuffers from the DOM API.
+type ExtendedSourceBuffer = SourceBuffer & {
+  ended: boolean
```
I think this needs to be optional, because it initially will be `undefined`

```suggestion
  ended?: boolean
```

itsjamie replied:
Good point. This must not have been caught by the TypeScript code that normally catches delayed initialization because the constructor isn't defined by TS.

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -219,9 +263,9 @@ class BufferController extends EventHandler {
-      timeRanges[streamType] = sourceBuffer[streamType].buffered;
+    const sbSet = this.sourceBuffer;
+    for (let streamType in sbSet) {
+      timeRanges[streamType] = sbSet[streamType as keyof SourceBufferSet].buffered;
```
`sbSet[streamType as keyof SourceBufferSet]` could be `null` according to the interface, which would be an error here

itsjamie replied:
Odd. At that section despite the interface saying it could be null, typescript is inferring the type as a pure ExtendedSourceBuffer. So for some reason it's left the null off.

I wonder why...

itsjamie replied:
Ah ha. Strict Null Checks has to be turned on. Enabling it and fixing these.

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -250,21 +305,25 @@ class BufferController extends EventHandler {
     let sourceBuffer = this.sourceBuffer;
     for (let type in sourceBuffer) {
-      let sb = sourceBuffer[type];
+      let sb = sourceBuffer[type as keyof SourceBufferSet];
```
Optional just means undefined, so it would still be a reference error. The only question is, if it is possible for the code to reach the caller with it being null or undefined. Otherwise you can "definitely" say its ok. With the `!`.

itsjamie replied:
Oh I see. But if we don't initialize the properties to null, then the `for in` won't find them and therefore it can't be null.

What does the hls.js project guarantee in the API each major version? If the controller public interfaces aren't in the guarantee I'm down to make this change, otherwise someone might be explicitly looking for `null` over `undefined`.

johnBartos replied:
Controller interfaces aren’t public; we can change them without a major bump

johnBartos commented:
Excellent work! 👍 thanks for taking point on the reviews @tjenkinson

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -328,13 +393,16 @@ class BufferController extends EventHandler {
+    let sbSet = this.sourceBuffer;
     let dataType = data.type;
-    for (let type in sb) {
+    for (let type in sbSet) {
```
Actually can't this whole loop construct be turned into:

```
const sb = sbSet[data.type]
if (!sb) {
  throw ...
}
... set ended
```

itsjamie replied:
It's part of the event bus call chain so the error will be caught there and turned into a hls.js error if sbSet[data.type] is undefined,

tjenkinson replied:
Yeh you're right there is actually no need for the loop. I'm not sure whether we should throw or not though as that would change the behaviour if there are cases now where this is called and for some reason the property is missing

itsjamie replied:
I took a second look, and noticed that it actually has some undocumented functionality that was being used in one place. If no type is passed on the data, it will actually clear all source buffers. So I kept the loop.

itsjamie replied:
Also added a comment describing that as intended functionality.

itsjamie commented on tsconfig.json:
```diff
@@ -4,7 +4,13 @@
     "target": "ES5",
-    "strict": false
+    "strict": false,
+    "strictNullChecks": true,
```
I think `strict: true` might be a good goal eventually, but I've spent some time trying to make the file meet strict guidelines, and I think it will be more than enough effort to just do a first-pass of applying typing and ensuring null checks across the rest of the codebase.

itsjamie replied:
I'll submit a commit with the changes to bring it to strict compliance so you can see though and I haven't wasted the effort for this file.

tjenkinson replied:
:+1: if it causes a load of work we should leave it as `false` and get this merged, and then follow up later :)

itsjamie replied:
A big part of the change that I wanted to leave out was that by enabling strict mode, it had a real problem with `index signatures`. That I think would need a major code change to work around.

itsjamie replied:
This would mean things like SourceBufferSet doesn't define a index signature, since it technically is an object and according to typescript should be accessed like `obj.property` rather than `obj[property]` without defining this index signature thing.

itsjamie commented:
So one aspect that I just fixed was that because of usage of things like `Object.keys` initializing the values to be `null` actually wasn't the same thing as allowing them to be optionally defined. This is now fixed with the above.

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -509,7 +583,9 @@ class BufferController extends EventHandler {
         return;
       }
       if (segments && segments.length) {
```
I agree. I did this, and swapped the logic to early exit since the whole block of code was wrapped in the conditional.

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -328,13 +393,16 @@ class BufferController extends EventHandler {
+        let sb = sbSet[type as keyof SourceBufferSet];
+        if (!sb) continue;
+
+        if (!sb.ended) {
```
Agreed. There was another place where there were multiple conditionals that would have to check if sb existed, so I left it this way there. And in the place where there was only the one, I put it on the same condition.

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -1,55 +1,102 @@
+ */
 
-import Event from '../events';
+import Events from '../events';
```
I had to rename the `Event` to `Events` because it conflicted with the typing for the default simple event for the DOM.

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -1,55 +1,102 @@
   }
 
-  onLevelPtsUpdated (data) {
+  onLevelPtsUpdated (data: any) {
```
looks simple to add a type for this
```suggestion
  onLevelPtsUpdated (data: { start: number }) {
```

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -151,28 +201,18 @@ class BufferController extends EventHandler {
       this.pendingTracks = {};
       this.tracks = {};
-      this.sourceBuffer = {};
+      this.sourceBuffer = { };
```
typo?

itsjamie replied:
Came from when I removed the `null` definitions I had in the middle because they didn't play nice with Object.keys.

itsjamie replied:
Fixed.

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -218,13 +270,17 @@ class BufferController extends EventHandler {
-    const sourceBuffer = this.sourceBuffer;
-    for (let streamType in sourceBuffer) {
-      timeRanges[streamType] = sourceBuffer[streamType].buffered;
+    let timeRanges: any = {};
```
can use `TimeRanges` here
```suggestion
    const timeRanges: { [streamType: keyof SourceBufferSet]: TimeRanges };
```

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -238,33 +294,37 @@ class BufferController extends EventHandler {
       }
     }
-    this.sourceBuffer = {};
+    this.sourceBuffer = { };
```
typo?

itsjamie replied:
```suggestion
    this.sourceBuffer = {};
```

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -484,86 +558,92 @@ class BufferController extends EventHandler {
+      return;
+    }
+
+    let segment = segments.shift();
```
could be `const`?

itsjamie replied:
```suggestion
    const segment = segments.shift();
```

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -394,20 +464,23 @@ class BufferController extends EventHandler {
 
-  onLevelUpdated ({ details }) {
-    if (details.fragments.length > 0) {
+  onLevelUpdated ({ details }: { details: any}) {
```
fragment looks like it might be a bit complicated and best left for now, but I think we could type the rest
```suggestion
  onLevelUpdated ({ details }: { details: { totalduration: number, targetduration?: number, averagetargetduration?: number, live: boolean, fragments: any[] }) {
```

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -151,28 +201,18 @@ class BufferController extends EventHandler {
       this.pendingTracks = {};
       this.tracks = {};
-      this.sourceBuffer = {};
+      this.sourceBuffer = { };
```
```suggestion
      this.sourceBuffer = {};
```

tjenkinson commented on karma.conf.js:
```diff
@@ -25,7 +25,7 @@ module.exports = function(config) {
 		// preprocess matching files before serving them to the browser
 		// available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
 		preprocessors: {
-		  '**/*.js': ['webpack', 'sourcemap']
```
Looks like there is now a combination of tabs and spaces in here

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -319,47 +386,44 @@ class BufferController extends EventHandler {
-          sb[type].ended = true;
-          logger.log(`${type} sourceBuffer now EOS`);
+  // an undefined data.type will mark all buffers as EOS.
+  onBufferEos (data: { type: SourceBufferName | undefined }) {
```
Could just do `type?:`

itsjamie replied:
👍

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -319,47 +386,44 @@ class BufferController extends EventHandler {
-          sb[type].ended = true;
-          logger.log(`${type} sourceBuffer now EOS`);
+  // an undefined data.type will mark all buffers as EOS.
+  onBufferEos (data: { type: SourceBufferName | undefined }) {
```
```suggestion
  onBufferEos (data: { type?: SourceBufferName }) {
```

tjenkinson commented on src/controller/buffer-controller.ts:
```diff
@@ -84,7 +133,7 @@ class BufferController extends EventHandler {
   }
 
-  onManifestParsed (data) {
+  onManifestParsed (data: any) {
```
```suggestion
  onManifestParsed (data: { altAudio: boolean }) {
```

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -394,20 +462,23 @@ class BufferController extends EventHandler {
-  onLevelUpdated ({ details }) {
-    if (details.fragments.length > 0) {
+  onLevelUpdated ({ details }: { details: { totalduration: number, targetduration?: number, averagetargetduration?: number, live: boolean, fragments: any[] } }) {
+    if (details.fragments.length <= 0) {
```
```suggestion
    if (details.fragments.length > 0) {
```

itsjamie commented on src/controller/buffer-controller.ts:
```diff
@@ -484,86 +556,92 @@ class BufferController extends EventHandler {
+          event.fatal = true;
+          hls.trigger(Events.ERROR, event);
+        } else {
+          hls.trigger(Events.ERROR, event);
```
I think so. I'll take a look.

itsjamie commented:
@tjenkinson or @michaelcunningham19 can we merge this if everything is good?

Thank yas.

tjenkinson commented:
:tada:
